libraries/Microsoft.Bot.Builder.Ai/TranslationMiddleware.cs

```diff
@@ -0,0 +1,177 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Linq;
+using System.Text.RegularExpressions;
+using System.Threading.Tasks;
+using Microsoft.Bot.Schema;
+
+namespace Microsoft.Bot.Builder.Ai
+{
+    public class TranslationMiddleware : IMiddleware
+    {
+        private readonly string[] _nativeLanguages;
+        private readonly Translator _translator;
+        private readonly Dictionary<string,List<string>> _patterns;
+        private readonly Func<ITurnContext, string> _getUserLanguage;
+        private readonly Func<ITurnContext, Task<bool>> _setUserLanguage;
+
+        /// <summary>
+        /// Constructor for automatic detection of user messages
+        /// </summary>
+        /// <param name="nativeLanguages">List of languages supported by your app</param>
+        /// <param name="translatorKey"></param>
+        public TranslationMiddleware(string[] nativeLanguages, string translatorKey)
+        {
+            if(nativeLanguages==null || nativeLanguages.Count()==0)
+                throw new ArgumentNullException(nameof(nativeLanguages));
+            this._nativeLanguages = nativeLanguages;
+            if(string.IsNullOrEmpty(translatorKey))
+                throw new ArgumentNullException(nameof(translatorKey));
+            this._translator = new Translator(translatorKey);
+            _patterns = new Dictionary<string, List<string>>();
+        }
+
+        /// <summary>
+        /// Constructor for automatic of user messages and using templates
+        /// </summary>
+        /// <param name="nativeLanguages">List of languages supported by your app</param>
+        /// <param name="translatorKey"></param>
+        /// <param name="patterns">Dictionary with language as a key and list of patterns as value</param>
+        public TranslationMiddleware(string[] nativeLanguages, string translatorKey, Dictionary<string, List<string>> patterns)
+        {
+            if (nativeLanguages == null || nativeLanguages.Count() == 0)
```

Comment:
Given how often you're making this check, consider making it a method to simplify all your code:

private static void AssertValidNativeLanguages(string [] nativeLanguages)
{
  if (nativeLanguages == null) throw new argumentNullException(nameof(...));
  if (nativeLanguages.Length == 0) throw new ArgumentException(nameof(...)); 
}

Reply:
Updated that would be better